### ISSUE TITLE
api root

### ISSUE THREAD

**Issue by Jcbobo88** [alanjds/drf-nested-routers#42]
nested resources not showed in api root

**Comment by alanjds** [comment]
Hi. Can you elaborate, with a minimal example please?

**Comment by Jcbobo88** [comment]
I will use the example in the docs

with 
from rest_framework_nested import routers
from views import DomainViewSet, NameserverViewSet
(...)

router = routers.DefaultRouter()
router.register(r'domains', DomainViewSet)

domains_router = routers.NestedSimpleRouter(router, r'domains', lookup='domain')
domains_router.register(r'nameservers', NameserverViewSet)

urlpatterns = patterns('',
    url(r'^', include(router.urls)),
    url(r'^', include(domains_router.urls)),
)

the routers are working, but when you go to the api-root view the view NameserverViewSet is not showed in the browsable api

I hope now is clear (sorry for the really poor detail)

**Comment by jsenecal** [comment]
@Jcbobo88 How should the default router display the nested URLs? They are not supposed to show up as the default router only shows the top level resources by design. Linking to those nested resources can be done by Modified HyperlinkRelatedFields and such.

**Comment by alanjds** [comment]
Seems reasonable for me to close the issue. Do you agree @Jcbobo88  ?

**Comment by ryanhiebert** [comment]
I agree that the links to nested resources should be from elsewhere than the api root.

**Comment by Jcbobo88** [comment]
I think that having the nested resource listed in the api root can help developers that need to use that resources, (all api listed and described in a single point), if you think there is a better option is the same. I did not understand how access the lists of nested resources.

**Comment by ryanhiebert** [comment]
I see your perspective. The problem comes when you attempt to generate the links to those nested routes. Which context are they a part of?

I'm not sure that this is something that the auto generated too can handle automatically in a sane way. I think that you'll have to end up making that serializer and viewer manually.

Do you have a better answer as to how to make the links to the sub resources?

**Comment by Jcbobo88** [comment]
No i don't have,  I just come back from hospital some day ago and start to look the mail...
( no one of mine colleagues care about that aspect so they skip it)

**Comment by alanjds** [comment]
Calm down, dudes :)

I tend to prefer less info, than the whole API Tree on the 1st page. It could hurt the service if is too big, but maybe it can be a on-off switch...

**Comment by ryanhiebert** [comment]
I don't have a particular problem with having an option to have all the routes on the main router. I just don't see a way to get the links for the nested routes. If that was figured out, then I'd no longer have a problem.

Off the top of my head the only way I can think to do that would be to use URI templates instead of straight hyperlinks. I don't know if it's feasible to generate URI templates from regex routes. Because of that, it makes me think that it's likely going to have to be custom rather than auto-generated.

I agree with @alanjds preference of less info on the 1st page, but I don't mind exploring the option of more if someone wants to figure out how that could be done.

**Comment by nachopro** [comment]
Hi, I'm a new user of this module. I switched my DefaultRouter to SimpleRouter and now I'm getting an 404 error in the API Root.

How I can fix this issue? thanks!

**Comment by alanjds** [comment]
@nachopro I guess your issue is another. SimpleRouter does not include the API Root. See this docs: http://www.django-rest-framework.org/api-guide/routers/#defaultrouter

**Comment by ryanhiebert** [comment]
@nachopro : That's right, you need to be using this package's `NestedDefaultRouter` for the API root. Check out the example in the [README file](../tree/master/README.md).

**Comment by nachopro** [comment]
I missed this line: https://github.com/alanjds/drf-nested-routers/blob/master/rest_framework_nested/routers.py#L30

**Comment by alanjds** [comment]
@nachopro :) No problem.

**Comment by timstallmann** [comment]
Hi -- not sure if this is the same issue, but I'm having trouble with the DRF auto-generated docs and this package. The first nested route I have (using `NestedDefaultRouter`) shows up in the docs, but others don't.. any suggestions?

**Comment by keithhackbarth** [comment]
Was this ever resolved? Seems like discussion ended at release 91, but I don't see what the resolution is.

I personally think the option of showing the nested APIs in the API root would be beneficial for developers.

**Comment by alanjds** [comment]
> I agree with @alanjds preference of less info on the 1st page, but I don't mind exploring the option of more if someone wants to figure out how that could be done.

As @ryanhiebert said, nobody opposed to it, but nobody stepped in and coded it.

Anyway, I would be happy with such PR if it have an on-off switch.

**Comment by gabn88** [comment]
Could we maybe use "actions" for it? To show possible child relations within the actions dropdown? 
The actions button/functionality has since 2018 got a major update, so now might be the time for it.

**Comment by kevinrenskers** [comment]
I ran into the same issue: wanting nested endpoints to be visible in the browsable API, and I did indeed solve it using actions.

```
class PostViewSet(viewsets.ModelViewSet):
    @action(detail=True, methods=["get"])
    def comments(self, request, pk: int):
        # This is a fake action so it shows up in the browsable API.
        # The real endpoint is handled by CommentViewSet.
        pass

class CommentViewSet(viewsets.ModelViewSet):
    serializer_class = serializers.CommentSerializer

    def get_queryset(self):
        post_id = self.kwargs["post_pk"]
        return Comment.objects.prefetch().filter(post_id=post_id)
```

urls.py:

```
from django.urls import include, path
from rest_framework.routers import DefaultRouter
from rest_framework_nested.routers import NestedDefaultRouter

from . import views

router = DefaultRouter(use_regex_path=False)
router.register("posts", views.PostViewSet, basename="post")

# Nested router for comments
post_router = NestedDefaultRouter(router, r"posts", lookup="post", use_regex_path=False)
post_router.register(r"comments", views.CommentViewSet, basename="post-comments")

urlpatterns = [
    path("", include(post_router.urls)),
    path("", include(router.urls)),
]
```

Now when someone uses the browsable API to look at the `/posts/{post_id}` endpoint there's the "Extra Actions" dropdown with Comments. 

The trick is to include the nested router urls first, and then the base router. If you do it the other way around then the action is used instead of the nested viewset.
